pkg/sql/bulkmerge/merge_processor.go

```diff
@@ -156,8 +165,117 @@ func (m *bulkMergeProcessor) handleRow(row rowenc.EncDatumRow) (rowenc.EncDatumR
+	if task == 0 {
+		return nil, splits[0]
+	}
+	last := len(splits) - 1
+	if task == last {
+		return splits[task], nil
+	}
+	if task > last {
+		panic(errors.AssertionFailedf("bad"))
+	}
+	return splits[task-1], splits[task]
+}
+
+// This is stolen from roachpb/data.go where we check if spans overlap.
+func isOverlapping(startKey, endKey roachpb.Key, sst execinfrapb.BulkMergeSpec_SST) bool {
+	otherStartKey, otherEndKey := sst.StartKey, sst.EndKey
+	if len(endKey) == 0 && len(otherEndKey) == 0 {
+		return startKey.Equal(otherStartKey)
+	} else if len(endKey) == 0 {
+		return bytes.Compare(startKey, otherStartKey) >= 0 && bytes.Compare(startKey, otherEndKey) < 0
+	} else if len(otherEndKey) == 0 {
+		return bytes.Compare(otherStartKey, startKey) >= 0 && bytes.Compare(otherStartKey, endKey) < 0
+	}
+	return bytes.Compare(endKey, otherStartKey) > 0 && bytes.Compare(startKey, otherEndKey) < 0
+}
+
+func (m *bulkMergeProcessor) mergeSSTs(ctx context.Context, taskID int) error {
+	// TODO(annie): get task from coordinator (look into how flows w multiple processors do it)
+	startKey, endKey := getKeysForTask(taskID, m.spec.Splits)
+	var storeFiles []storageccl.StoreFile
+	outputName := fmt.Sprintf("%s%d.sst", *m.spec.OutputUri, taskID)
+	dest, err := m.flowCtx.Cfg.ExternalStorageFromURI(ctx, outputName, username.RootUserName())
+	if err != nil {
+		return err
+	}
+	// Find the ssts that overlap with the given task's key range.
+	for _, sst := range m.spec.Ssts {
+		var store cloud.ExternalStorage
+		uri, err := url.Parse(sst.Uri)
+		if err != nil {
+			return err
+		}
+		// TODO(annie): cache
+		store, err = m.flowCtx.Cfg.ExternalStorageFromURI(ctx, sst.Uri, username.RootUserName())
+		if err != nil {
+			return err
+		}
+
+		if isOverlapping(startKey, endKey, sst) {
+			// TODO(annie): check if this is the correct filePath
+			storeFiles = append(storeFiles, storageccl.StoreFile{Store: store, FilePath: uri.Path})
+		}
+
+		iterOpts := storage.IterOptions{
+			KeyTypes:   storage.IterKeyTypePointsAndRanges,
+			LowerBound: startKey,
+			UpperBound: endKey,
+		}
+		iter, err := storageccl.ExternalSSTReader(ctx, storeFiles, nil, iterOpts)
+		if err != nil {
+			return err
+		}
+		defer iter.Close()
+		// This uses bulkio.backup.file_size: which is 128MB by default.
+		sinkConf := backupsink.SSTSinkConf{
+			ID:       m.flowCtx.NodeID.SQLInstanceID(),
+			Enc:      nil,
+			ProgCh:   nil,
+			Settings: &m.flowCtx.Cfg.Settings.SV,
+		}
+		for iter.SeekGE(storage.MVCCKey{Key: sst.StartKey}); ; iter.NextKey() {
+			ok, err := iter.Valid()
+			if err != nil {
+				return err
+			}
+			if !ok {
+				break
+			}
+			key := iter.UnsafeKey()
+			val, err := iter.UnsafeValue()
+			if err != nil {
+				return err
+			}
+			writer, err := backupsink.MakeSSTSinkKeyWriter(sinkConf, dest, nil)
+			if err != nil {
+				return err
+			}
+			_, err = writer.Write(ctx, backupsink.ExportedSpan{
```

Comment:
I think this should be calling WriteKey instead of Write. `Write` looks like it expects a pre-formatted SST.